mac/Keyman4MacIM/Keyman4MacIM/KeySender.m

```diff
@@ -62,20 +56,54 @@ - (void)sendKeyDown:(NSUInteger)keyCode forSourceEvent:(NSEvent *)event includeK
   }
 }
 
--(void) dealloc {
-  CFRelease(self.eventSource);
+- (void)sendBackspaceforEventSource:(CGEventSourceRef)eventSource {
+  [self.appDelegate logDebugMessage:@"KeySender sendBackspaceforSourceEvent"];
+
+  [self postKeyboardEventWithSource:eventSource code:kVK_Delete postCallback:^(CGEventRef eventToPost) {
+      CGEventPost(kCGHIDEventTap, eventToPost);
+  }];
 }
 
-- (void)sendBackspaceforSourceEvent:(NSEvent *)event {
-  [self.appDelegate logDebugMessage:@"KeySender sendBackspaceforSourceEvent"];
-  [self sendKeyDown:kVK_Delete forSourceEvent:event includeKeyUp:YES];
+- (void)postKeyboardEventWithSource: (CGEventSourceRef)source code:(CGKeyCode) virtualKey postCallback:(PostEventCallback)postEvent{
+
+    CGEventRef ev = CGEventCreateKeyboardEvent (source, virtualKey, true); //down
+    if (postEvent)
+        postEvent(ev);
+    CFRelease(ev);
+    ev = CGEventCreateKeyboardEvent (source, virtualKey, false); //up
+    if (postEvent)
+        postEvent(ev);
+    CFRelease(ev);
 }
 
+/**
+ sendKeymanKeyCodeForEvent sends the kKeymanEventKeyCode to the
+ frontmost application to indicate that all the backspaces have been processed
+ and we can insert the queuedText to the client
+ */
+
 - (void)sendKeymanKeyCodeForEvent:(NSEvent *)event {
   [self.appDelegate logDebugMessage:@"KeySender sendKeymanKeyCodeForEvent"];
-  // this is not a real keycode, so we do not need a key up event
-  // kKeymanEventKeyCode is used to indicate that all the backspaces have been processed
-  // and we can insert the queuedText to the client
+  
   [self sendKeyDown:kKeymanEventKeyCode forSourceEvent:event includeKeyUp:NO];
+  
+  // Returns the frontmost app, which is the app that receives key events.
+  NSRunningApplication *app = NSWorkspace.sharedWorkspace.frontmostApplication;
+  pid_t processId = app.processIdentifier;
+  NSString *appName = app.localizedName;
```

Comment:
is `localizedName` the most helpful string for debug logging? Might not be better to use an identifier of some kind?

Reply:
changed to use bundleId